### PR TITLE
GH-1265: Document `How to...` use SpEL-functions

### PR DESCRIPTION
**Pull request by artembilan**
Fixes spring-cloud/spring-cloud-dataflow#1265

### CODE REVIEW

File: spring-cloud-dataflow-docs/src/main/asciidoc/howto.adoc [hunk 1]
```diff
@@ -170,3 +170,31 @@ These properties can also be specified via `deployment` properties when deployin
 ----
 dataflow:>stream deploy foo --properties "app.*.logging.level.org.springframework.integration=DEBUG"
 ----
+
+=== Advanced SpEL expressions with stream
+
+One of the powerful feature of SpEL expressions is http://docs.spring.io/spring/docs/current/spring-framework-reference/html/expressions.html#expressions-ref-functions[functions].
```
Comment (ilayaperumalg) [inline spring-cloud-dataflow-docs/src/main/asciidoc/howto.adoc]: typo: feature -> features

File: spring-cloud-dataflow-docs/src/main/asciidoc/howto.adoc [hunk 2]
```diff
@@ -170,3 +170,31 @@ These properties can also be specified via `deployment` properties when deployin
...
+Spring Integration provides `jsonPath()` and `xpath()` out-of-the-box http://docs.spring.io/spring-integration/reference/html/spel.html#spel-functions[SpEL-functions], if appropriate libraries are in the classpath.
+All the provided Spring Cloud Stream application starters are supplied with the `json-path` and `spring-integration-xml` jars, thus we can use those SpEL-functions in Spring Cloud Data Flow streams whenever expressions are possible.
+For example we can transform JSON-aware `payload` from the HTTP request using some `jsonPath()` expression:
+
+[source,bash]
```
Comment (ericbottard) [inline spring-cloud-dataflow-docs/src/main/asciidoc/howto.adoc]: Not really `bash`

File: spring-cloud-dataflow-docs/src/main/asciidoc/howto.adoc [hunk 3]
```diff
@@ -170,3 +170,31 @@ These properties can also be specified via `deployment` properties when deployin
...
+----
+dataflow:>stream create jsonPathTransform --definition "http | transform --expression=#jsonPath(payload,'$.price') | log" --deploy
+...
+dataflow:> http post --target http://localhost:8080 --data {"symbol":"SCDF","price":72.04}
+dataflow:> http post --target http://localhost:8080 --data {"symbol":"SCDF","price":72.06}
+dataflow:> http post --target http://localhost:8080 --data {"symbol":"SCDF","price":72.08}
+----
+In this sample we apply jsonPath for the incoming payload to extract just only `price` field value.
```
Comment (ericbottard) [inline spring-cloud-dataflow-docs/src/main/asciidoc/howto.adoc]: typo: "the" `price` field value

File: spring-cloud-dataflow-docs/src/main/asciidoc/howto.adoc [hunk 4]
```diff
@@ -170,3 +170,31 @@ These properties can also be specified via `deployment` properties when deployin
...
+Similar syntax can be used with `splitter` or `filter` `expression` option.
```
Comment (ericbottard) [inline spring-cloud-dataflow-docs/src/main/asciidoc/howto.adoc]: I think optionS with an 's' would be more natural

File: spring-cloud-dataflow-docs/src/main/asciidoc/howto.adoc [hunk 5]
```diff
@@ -170,3 +170,31 @@ These properties can also be specified via `deployment` properties when deployin
...
+Actually any available SpEL-based option has access to the built-in SpEL-functions.
+For example we can extract some value from JSON data to calculate the `partitionKey` before sending output to the Binder:
+
+[source,bash]
+----
+dataflow:>stream deploy foo --properties "deployer.transform.count=2,app.transform.producer.partitionKeyExpression=#jsonPath(payload,'$.symbol')"
+----
+The same syntax can be applied for `xpath()` SpEL-function when you deal with XML data.
+Any other custom SpEL-function also can be used, but for this purpose you should build a library with the `@Configuration` class containing an appropriate `SpelFunctionFactoryBean` `@Bean` definition.
```
Comment (ericbottard) [inline spring-cloud-dataflow-docs/src/main/asciidoc/howto.adoc]: can <-> also be used

File: spring-cloud-dataflow-docs/src/main/asciidoc/howto.adoc [hunk 6]
```diff
@@ -170,3 +170,31 @@ These properties can also be specified via `deployment` properties when deployin
...
+The target Spring Cloud Stream application starter should be re-packaged to supply such a custom extension via build-in Spring Boot `@ComponentScan` mechanism.
```
Comment (ericbottard) [inline spring-cloud-dataflow-docs/src/main/asciidoc/howto.adoc]: builT-in.

But should we actually mention `@ComponentScan` (as opposed to say, auto-configuration, or something more explicit and deemed less "dangerous")

File: spring-cloud-dataflow-docs/src/main/asciidoc/howto.adoc (hunk 2 above)
Comment (artembilan) [inline spring-cloud-dataflow-docs/src/main/asciidoc/howto.adoc]: Well, that is copy/paste from other similar places in this Reference Manual.
For example `streams.adoc`
Not my fault 😄

Comment (markpollack) [inline spring-cloud-dataflow-docs/src/main/asciidoc/howto.adoc]: I think it helps w/ the color formatting.

**Comment by markpollack** [comment]
squashed and merged as eacff82e0e942244ae1ebc9b2a7bb16af7fea2fa

**Pull request closed without merging**